### PR TITLE
Implement (thread-safe) signal handlers for non-introspected object

### PR DESCRIPTION
**Pull request by wotzlaff**
The title says almost everything. The proposed changes are quite elaborate as I stumbled upon another problem while implementing non-gi signals: It is possible (and in GStreamer it's not unusal) that signal handlers are called from another thread. Then, some locking mechanism has to be used in order to prevent a segfault (which is caused by using the content of a `Persistent` from a thread which is different from thread of the uv main loop).

There are three major parts:
- I stripped down `SignalConnect` in `gobject.cc`: Signals are handled at runtime. We actually do not need the argument information which is provided from GIR. I'd propose to use `GValueToV8` instead of `GIArgumentToV8`.
- I extracted the calling code from `Closure::Marshal` to a new function `Closure::Execute`. Functionality does not change when the function was called from the main thread. Otherwise, the call will be serialized into some `CallbackWrapper` which is used later on in the main thread.
- I transferred the locking mechanism from [here](https://github.com/node-ffi-napi/node-ffi-napi). The downside is that one has to use some low-level libuv routines. This could cause trouble. One should also note that the js callback is only if the main loop is not busy, but we won't get around this issue. In particular, it's quite easy to get a deadlock.

Be aware that I'm still no expert in GIR or libuv. So maybe there's something terribly wrong with my proposal...

This partially solves my issue #183.

### CODE REVIEW

**Comment by romgrk** [comment]
Do you have any file I can test this with? I'll give you feedback soon :)

**Comment by romgrk** [comment]
Btw, one comment about naming conventions. This project is a bit confused, but it's not it's fault, because it's stuck between `glib` that uses `c_style_naming_conventions` and `V8` that uses `Cpp.NamingConventions` :/

Usually I use `snake_case` for variables that hold `glib` stuff, and `camelCase` for variables that hold `V8` stuff. I don't really know what to do with all that, but if you wanna use `camelCase` only that's fine with me.

File: src/closure.cc [hunk 1]
```diff
@@ -18,88 +18,153 @@ using Nan::Persistent;
...
+
+        uv_mutex_lock(&env->mutex);
+        env->queue.push(cb);
+        uv_mutex_unlock(&env->mutex);
+        uv_async_send(&Closure::asyncHandle);
+
+        cb->Wait();
+        delete cb;
+    }
+}
```
[the remark this answers is not in the document]
Comment (wotzlaff) [reply src/closure.cc:96]: Yes, you're right. Thanks for the patch. Nice that you also found that `cb` could be stack allocated. I did `delete` it at another place which was terribly wrong and forget to check allocation here.

About the first argument in `param_values`: Of course there's mixing between the logics. It was just a quick idea to get rid of the first argument here because I didn't want to copy it in `Prepare`.

File: src/closure.cc [hunk 2]
```diff
@@ -18,88 +18,153 @@ using Nan::Persistent;
...
+    auto& handle = Closure::asyncHandle;
+    AsyncCallEnvironment* env = new AsyncCallEnvironment();
+    handle.data = env;
+    env->mainThread = uv_thread_self();
+    uv_loop_t* loop = uv_default_loop();
+    uv_async_init(loop, &handle, Closure::QueueHandler);
+    uv_mutex_init(&env->mutex);
+    uv_unref(reinterpret_cast<uv_handle_t *>(&handle));
+    uv_async_send(&handle);
+}
```
[the remark this answers is not in the document]
Comment (wotzlaff) [reply src/closure.cc:129]: Did you mean the whole function `Initialize` or just these three lines?

File: src/gobject.cc [hunk 3]
```diff
@@ -250,36 +250,20 @@ NAN_METHOD(SignalConnect) {
...
-    const char *signalName = *Nan::Utf8String (TO_STRING (info[0]));
-    GISignalInfo *signal_info = FindSignalInfo (object_info, signalName);
-
-    if (signal_info == NULL) {
-        Throw::SignalNotFound(object_info, signalName);
-        goto out;
+    const char* signalName = *Nan::Utf8String(TO_STRING(info[0]));
+    if (!g_signal_parse_name(signalName, gtype, &signalId, &detail, FALSE)) {
+        Nan::ThrowTypeError("Signal name is invalid");
+        return;
```
[the remark this answers is not in the document]
Comment (wotzlaff) [reply src/gobject.cc:271]: Yeah... It's really not an `ThrowTypeError`.
Currently I'm only parsing the signal name which is not enough. I should try to find out whether the object can have the signal and throw if this is not the case. The current solution is too easy.

**Comment by wotzlaff** [comment]
I have been already wondering about the naming convention. As I'm more familiar with JS than C, I'm using the camelCase version. But feel free to ask for changes when this is not suitable.

BTW: I messed up with my repo using a force push as I didn't get the right syntax for rebase. Hopefully you'll still be able to find everything which is needed to comment the PR.

File: src/closure.cc [hunk 4]
```diff
@@ -18,88 +18,153 @@ using Nan::Persistent;
 
 namespace GNodeJS {
 
-GClosure *Closure::New (Local<Function> function, GICallableInfo* info, guint signalId) {
-    Closure *closure = (Closure *) g_closure_new_simple (sizeof (*closure), GUINT_TO_POINTER(signalId));
+GClosure *Closure::New(Local<Function> function, guint signalId) {
+    Closure *closure = (Closure *) g_closure_new_simple (sizeof(Closure), NULL);
```
[the remark this answers is not in the document]
Comment (wotzlaff) [reply src/closure.cc]: I don't fully understand why we need it. Probably it has something to do with the interference of `mustCopy`?
You can have it back :)

File: src/gobject.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (wotzlaff) [reply src/gobject.cc:272]: Should we also throw if `g_signal_connect_closure` returns 0?

**Comment by wotzlaff** [comment]
As you've already seen, testing is not a passion of mine. I tried to implement a test to handle the case of function calling from another thread and non-introspected objects. Probably these should be two independent test. But the same could be said about mostly all test which we're using here...

File: src/gobject.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (romgrk) [reply src/gobject.cc:273]: Na, we better handle it in https://github.com/romgrk/node-gtk/blob/ddd1ba2135e0169376c8db28ce9ad37543e0aeac/lib/index.js#L57

File: src/closure.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (romgrk) [reply src/closure.cc]: Lol yeah give it back to me :)
Yes it's for the `g_signal_query` that returns additional info on the lifetime of some objects. In most cases though it woud be fine to just always copy, as long as we clear it after.

File: src/closure.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (romgrk) [reply src/closure.cc:152]: Oh yes I see, I hadn't read that part yet when I made this comment. I guess it can be removed before the call dispatch, just make sure to separate the js arguments preparation and the call dispatch.

File: src/closure.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (romgrk) [reply src/closure.cc]: The Initialize function

**Comment by romgrk** [comment]
> As you've already seen, testing is not a passion of mine.

I used to be like that, but TBH this project is what made me fall in love with tests. They don't make much sense eg when you develop a front-end app because interface are less testable and requirements move more, but for long-term projects that have well defined requirements and many tricky edge-cases (exactly like node-gtk), tests give you some much peace of mind when you make changes. The more (good) tests you have, the more sure you are to never break anything. That's why I've been pushing to have tests for each new feature, those are features we're less likely to break :) I hope this has contributed to giving you some of my passion for tests.

I've left a few comments but I haven't tested the new changes yet, I'll try to do that today or tomorrow.

**Comment by wotzlaff** [comment]
I like the idea to have a fully tested library. The problem is that there are too many untested execution paths in the current state. I'll try to provide tests for the things I'm implementing but it's quite hard for me to find functions in the GI repository which cover exactly the thing I want to point out. Is there any tool to query the functions in GIR?

Below is a stack trace for the segfault happening under macOS. It seems that the problems comes from `emit`thing the `GdkKeyEvent`. Probably this failure is not due to my changes. Maybe you have an idea.
```
  * frame #0: 0x00007fff6c91de5d libobjc.A.dylib`objc_msgSend + 29
    frame #1: 0x000000010b0f71d1 im-quartz.so`quartz_filter_keypress + 145
    frame #2: 0x000000010786285c libgtk-3.0.dylib`gtk_entry_key_press + 220
    frame #3: 0x00000001077798d3 libgtk-3.0.dylib`_gtk_marshal_BOOLEAN__BOXED + 131
    frame #4: 0x000000010475b79d libgobject-2.0.0.dylib`g_closure_invoke + 189
    frame #5: 0x0000000104771dcd libgobject-2.0.0.dylib`signal_emit_unlocked_R + 1965
    frame #6: 0x0000000104771543 libgobject-2.0.0.dylib`g_signal_emitv + 1107
    frame #7: 0x0000000104663f09 node_gtk.node`GNodeJS::SignalEmit(info=0x00007ffeefbfd900) at gobject.cc:380:9
    frame #8: 0x0000000104657249 node_gtk.node`Nan::imp::FunctionCallbackWrapper(info=0x00007ffeefbfd9e0) at nan_callbacks_12_inl.h:176:3
    frame #9: 0x00000001001d143e node`v8::internal::FunctionCallbackArguments::Call(v8::internal::CallHandlerInfo) + 524
    frame #10: 0x00000001001d0ae2 node`v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, v8::internal::BuiltinArguments) + 510
    frame #11: 0x00000001001d0318 node`v8::internal::Builtin_Impl_HandleApiCall(v8::internal::BuiltinArguments, v8::internal::Isolate*) + 252
...
```

**Comment by wotzlaff** [comment]
I tried to restructure everything a bit. Now, the responsibilities of each class should be a bit more clear. Feel free to comment if I misunderstood something or missed relationship.
The split is also necessary to point out the reusable parts in order to tackle the threading issues for callbacks later on.

**Comment by romgrk** [comment]
I'm too busy with work stuff :/ I'll try to review by the end of the week.

File: src/closure.cc [hunk 5]
```diff
@@ -51,22 +52,36 @@ void Closure::Marshal(GClosure     *base,
...
-        bool mustCopy = true;
+            if (signal_query.signal_id) {
+                mustCopy = (signal_query.param_types[i - 1] & G_SIGNAL_TYPE_STATIC_SCOPE) == 0;
+            }
 
-        if (signal_query.signal_id) {
-            mustCopy = (signal_query.param_types[i - 1] & G_SIGNAL_TYPE_STATIC_SCOPE) == 0;
+            js_args[i - 1] = GIArgumentToV8(&type_info, &argument, -1, mustCopy);
+        }
+        g_base_info_unref(info);
```
Comment (romgrk) [inline src/closure.cc:74]: The `g_base_info_(un)ref(info)` calls are unnecessary here.

File: src/closure.cc [hunk 6]
```diff
@@ -97,6 +112,21 @@ void Closure::Marshal(GClosure     *base,
...
+                      gpointer      marshal_data) {
+
+    auto closure = (Closure *) base;
+    auto signal_id = GPOINTER_TO_UINT(marshal_data);
+
+    AsyncCallEnvironment* env = reinterpret_cast<AsyncCallEnvironment *>(AsyncCallEnvironment::asyncHandle.data);
+    env->Call([&]() {
+        Closure::Execute(closure->info, signal_id, closure->persistent, g_return_value, n_param_values, param_values);
+    });
+}
```
Comment (romgrk) [inline src/closure.cc:129]: Would you mind explaining why it's now always wrapped in an `AsyncCallEnvironment::Call` instead of directly calling when it's in the same thread?

Comment (wotzlaff) [reply src/closure.cc:138]: I thought it would good to avoid code duplication. But it is really not necessary to wrap the call into a function when we are on the same thread. So I rewrote it a bit.

**Comment by romgrk** [comment]
LGTM. The macOS crashes are now present in master as well, they're probably caused by something else. Something similar happened a few months back, disappeared by itself at some point. I'll merge as it is, if there are adjustments to make we'll do them in time.

Thanks for your work :)

Btw, forgot to answer your question about inspecting GIR: check `./lib/inspect.js`, use it like this:
```js
const i = require('./lib/inspect.js')
const gtk = i.parseNamespace('Gtk', '3.0')
console.log(gtk)
console.log(i.infos)
```
